File: core/src/main/java/com/xuxue/spider/core/task/Task.java
package com.xuxue.spider.core.task;

import com.xuxue.spider.core.util.Serializabled;
import org.apache.http.client.methods.HttpUriRequest;

import java.io.*;
import java.net.URI;

/**
 * 这个接口定义了一个任务的规范
 *
 * 每一个任务必须能生成一个HttpRequest  必须能返回这个任务的URL的host
 * 必须定义了Cookie的等级
 *
 * @author  xuxue
 * @time 2016-6-13
 */
public interface Task extends Serializabled {

    /**
     *一个任务必须能够生成这个任务对应的Http请求
     * @return
     */
    HttpUriRequest buildHttpRequest();

    /**
     * 返回这个任务对应的Http请求地址的主机名
     * @return
     */
    String getHost();

    /**
     * 返回这个Task的URL
     * @return
     */
    URI getURI();

    /**
     * 返回这个任务需求的Cookie的等级
     * @return
     */
    CookieLeave getCookieLeave();

    /**
     * 返回这个任务的名字
     * @return
     */
    String getName();

    /**
     * 返回这个任务的id
     * @return
     */
    int getID();


    static <T> T toTask(byte[] data)throws IOException,ClassNotFoundException{
        ByteArrayInputStream in=new ByteArrayInputStream(data);
        ObjectInputStream obin=new ObjectInputStream(in);
        Object obj=obin.readObject();
        return (T)obj;
    }

}

File: core/src/main/scala/com/xuxue/spider/core/scala/task/Task.scala
package com.xuxue.spider.core.scala.task

import com.xuxue.spider.core.scala.util.Writeable
import java.net.URI

import com.xuxue.spider.core.scala.task.CookieLeave.CookieLeave
import org.apache.http.HttpHost
import org.apache.http.client.config.RequestConfig
import org.apache.http.client.methods.HttpUriRequest

import scala.util.Try
/**
  * Created by HanHan on 2016/7/7.
  */
trait Task extends Writeable{

  val connectionTimeout:Int

  val readTimeout:Int

  val uri:URI

  val name:String

  val cookieLeave:CookieLeave

  val maxDirection:Int

  val id:Int

  val size=toBytes().get.length

  def buildURIRequest(httpHost:HttpHost):Try[HttpUriRequest]

  def buildURIRequest():Try[HttpUriRequest]={
    buildURIRequest(null)
  }

  def host:String=uri.getHost

  def buildRequestConfig(httpHost:HttpHost):RequestConfig={
    RequestConfig.custom()
      .setConnectionRequestTimeout(connectionTimeout)
      .setConnectionRequestTimeout(connectionTimeout)
      .setSocketTimeout(readTimeout)
      .setMaxRedirects(maxDirection)
      .build()
  }
  override def toString={
    s"name=$name id=$id uri=$uri cookieLeave=$cookieLeave " +
      s"connectionTimeout=$connectionTimeout readTimeout=$readTimeout" +
      s"maxDirection=$maxDirection"
  }
}
